src/provider/store/Store.ts

```diff
@@ -1,122 +1,92 @@
-import {RootState} from './State';
-
-export type StoreChangeObserver<T> = (oldValue: T, newValue: T) => void;
 
 /**
  * Subset of properties of `Store` that are safe for use from actions and components.
  */
-export interface StoreAPI {
-    state: RootState;
-    dispatch(action: RootAction): Promise<void>;
+export interface StoreAPI<S, A> {
+    state: S;
+    dispatch(action: A): Promise<void>;
 }
 
-@injectable()
-export class Store extends AsyncInit implements StoreAPI {
-    private static INITIAL_STATE: RootState = {
-        notifications: [],
-        windowVisible: false
-    };
+export abstract class Action<S> {
+    public readonly type?: string;
 
-    public readonly onAction: Signal<[RootAction], Promise<void>> = new Signal(Aggregators.AWAIT_VOID);
+    constructor(type?: string) {
+        this.type = type;
+    }
 
-    private readonly _database: Database;
-    private _actionHandlerMap: ActionHandlerMap;
-    private _store!: ReduxStore<RootState, RootAction>;
+    public reduce(state: S): S {
+        return state;
+    }
+}
 
-    constructor(@inject(Inject.ACTION_HANDLER_MAP) actionHandlerMap: ActionHandlerMap, @inject(Inject.DATABASE) database: Database) {
-        super();
+export abstract class AsyncAction<S> extends Action<S> {
+    public abstract async dispatch(api: StoreAPI<S, Action<S>>): Promise<void>;
+}
+
+export interface Reducer<S> {
+    reduce(state: S): S;
+}
+
+type Listener<S> = (getState: () => S) => void;
+
+export class Store<S, A extends Action<S>> extends AsyncInit implements StoreAPI<S, A> {
+    private _currentState: S;
+    private _listeners: Listener<S>[] = [];
+
+    public readonly onAction: Signal<[A], Promise<void>, Promise<void>> = new Signal(Aggregators.AWAIT_VOID);
 
-        this._database = database;
-        this._actionHandlerMap = actionHandlerMap;
+    constructor(initialState: S) {
+        super();
+        this._currentState = initialState;
     }
 
-    protected async init(): Promise<void> {
-        await this._database.initialized;
-        this._store = createStore<RootState, RootAction, {}, {}>(this.reduce.bind(this), await this.getInitialState(), this.createEnhancer());
+    public get state(): S {
+        return this._currentState;
     }
 
-    public get state(): RootState {
-        return this._store.getState() as RootState;
+    public getState(): S {
+        return this._currentState;
     }
 
-    public async dispatch(action: RootAction): Promise<void> {
-        if (action instanceof CustomAction) {
+    protected async init(): Promise<void> {}
+
+    public async dispatch(action: A): Promise<void> {
+        if (action instanceof AsyncAction) {
             // Action has custom dispatch logic
             await action.dispatch(this);
-        } else {
-            // Pass straight through to redux store
-            await this._store.dispatch({...action});
         }
+        this.reduceAndEmit(action);
```

Comment:
Should this be `await`-ed?

Reply:
I don't think that's necessery at that stage as that's the final thing in the final thing in the store pipeline

Reply:
Then reduceAndSignal shouldn't need to be async and the await on `this.onAction.emit(action);`, shouldn't be necessary, right?

Reply:
sorry, you're right but we want to transfer that option to caller of dispatch.

Reply:
Then do you want `return this.reduceAndSignal(action);`?

Like, when do we want the promise returned by `dispatch(...)` to resolve? After all the listeners on `onAction` have resolved? Or before then?

Reply:
the dispatch call will return a promise to signal.emit down the line (reduce is synchronous) or await dispatch will reduce and then await the emit.